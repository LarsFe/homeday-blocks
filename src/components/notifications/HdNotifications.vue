<template>
  <div class="notifications">
    <transition-group
      name="notifications-bar-"
      tag="div"
    >
      <HdNotificationsBar
        v-for="(notification, i) in notifications"
        ref="notifications"
        :key="notification.id"
        :type="notification.type"
        :message="notification.message"
        :custom-icon="notification.customIcon"
        :visible="i === notifications.length - 1"
        v-bind="$attrs"
      >
        <slot :notification="notification" />
      </HdNotificationsBar>
    </transition-group>
    <div
      ref="sizer"
      :style="{
        height: `${sizerHeight}px`,
      }"
      class="notifications__sizer"
    />
  </div>
</template>

<script>
import onResize from 'homeday-blocks/src/services/on-resize';
import HdNotificationsBar from 'homeday-blocks/src/components/notifications/HdNotificationsBar.vue';

export default {
  name: 'HdNotifications',
  components: {
    HdNotificationsBar,
  },
  inheritAttrs: false,
  props: {
    notifications: {
      type: Array,
      default: () => [],
    },
  },
  data() {
    return {
      sizerHeight: 0,
    };
  },
  watch: {
    notifications() {
      this.$nextTick(this.resizeNotifications);
    },
    sizerHeight() {
      this.$emit('heightChange', this.sizerHeight);
    },
  },
  mounted() {
    this.resizeNotifications();
    this.addResizeEvents();
    this.addRoutingEvents();
  },
  beforeDestroy() {
    this.removeResizeEvents();
    this.removeRoutingEvents();
  },
  methods: {
    // Encapsulated in a method to be able to mock it in the tests
    getScrollHeight(el) {
      return el.scrollHeight;
    },
    resizeNotifications() {
      if (!this.notifications.length) {
        this.sizerHeight = 0;
        return;
      }

      if (!this.$refs.notifications) {
        this.sizerHeight = 0;
        return;
      }
      // Go through all the notifications and get the max height in order to
      // resize the sizer (since notification bars are position: fixed)
      const maxSize = this.$refs.notifications.reduce((maxHeight, { $el }) => {
        const elHeight = this.getScrollHeight($el);

        if (maxHeight >= elHeight) {
          return maxHeight;
        }

        return elHeight;
      }, 0);
      this.sizerHeight = maxSize;
    },
    addRoutingEvents() {
      this.$el.addEventListener('click', this.routeOnClick, false);
    },
    // We are catching cliks on anchors, and if it's an internal link, we use
    // router push instead (for a nicer transition)
    routeOnClick(e) {
      if (e.target.nodeName !== 'A') {
        return;
      }

      if (e.target.hostname !== window.location.hostname) {
        return;
      }

      e.preventDefault();
      this.$emit('route', { path: e.target.pathname });
    },
    resizeHandler() {
      this.resizeNotifications();
    },
    addResizeEvents() {
      onResize.onDebounced(this.resizeHandler);
    },
    removeResizeEvents() {
      onResize.offDebounced(this.resizeHandler);
    },
    removeRoutingEvents() {
      this.$el.removeEventListener('click', this.routeOnClick, false);
    },
  },
};
</script>

<style lang="scss">
@import 'homeday-blocks/src/styles/mixins.scss';

.notifications__sizer {
  transition: height ($time-s * 2) ease-in-out;
}
</style>
