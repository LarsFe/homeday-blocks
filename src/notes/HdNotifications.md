Sample Vuex store module:

```js
import uuidv4 from 'uuid/v4';
import Vue from 'vue';

const NOTIFICATION_DEFAULTS = {
  type: 'notification',
  message: '',
  duration: 5000,
  customIcon: '',
};

export default {
  state: {
    notifications: [],
  },
  actions: {
    addNotification({ commit, dispatch, state }, {
      type = NOTIFICATION_DEFAULTS.type,
      message = NOTIFICATION_DEFAULTS.message,
      customIcon = NOTIFICATION_DEFAULTS.customIcon,
      // For how long should the notification be displayed before it hides itself
      // Set to 0 for a permanent notification
      duration = NOTIFICATION_DEFAULTS.duration,
    }) {
      const notification = {
        type,
        message,
        customIcon,
        id: uuidv4(),
      };

      commit('setNotifications', [...state.notifications, notification]);

      if (duration > 0) {
        setTimeout(() => {
          dispatch('removeNotification', {
            id: notification.id,
          });
        }, duration);
      }

      return notification.id;
    },
    removeNotification({ commit, state }, { id: idToRemove }) {
      const filteredNotifications = state.notifications.filter(({ id }) => id !== idToRemove);
      commit('setNotifications', filteredNotifications);
    },
  },
  mutations: {
    setNotifications(state, notifications) {
      Vue.set(state, 'notifications', notifications);
    },
  },
};
```