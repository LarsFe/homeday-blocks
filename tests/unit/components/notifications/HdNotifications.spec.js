import { mount } from '@vue/test-utils';
import Vue from 'vue';
import HdNotifications from '@/components/notifications/HdNotifications.vue';

const NOTIFICATIONS = [
  {
    message: 'Hello <a href="http://localhost">Homeday!</a>',
  },
];

describe('HdNotifications', () => {
  let wrapper;
  let scrollHeightSpy = jest.fn();

  beforeEach(() => {
    wrapper = mount(HdNotifications, {
      propsData: {
        notifications: [],
      },
    });
  });

  afterEach(() => {
    scrollHeightSpy.mockRestore();
  });

  test('renders component', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('fires the heightChange event when height of the notifications changes', () => {
    scrollHeightSpy = jest.spyOn(wrapper.vm, 'getScrollHeight')
      .mockImplementation(
        // A random number, used just to make sure it's not 0 and that the height
        // actually changes
        () => 123,
      );

    wrapper.setProps({ notifications: NOTIFICATIONS });

    return Vue.nextTick()
      .then(() => {
        expect(wrapper.emitted('heightChange')).toBeTruthy();
      });
  });

  test('fires the route event on anchor click', () => {
    wrapper.setProps({ notifications: NOTIFICATIONS });

    const anchor = wrapper.find('a');

    anchor.trigger('click');
    expect(wrapper.emitted('route')).toBeTruthy();
    expect(wrapper.emitted('route')[0][0].path).toEqual(anchor.element.pathname);
  });
});
