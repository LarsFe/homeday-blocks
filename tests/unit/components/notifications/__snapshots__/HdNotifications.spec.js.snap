// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdNotifications renders the component corrently 1`] = `
<div class="notifications"><span name="notifications-bar-" tag="div"><div class="notifications-bar notifications-bar--notification"><div class="notifications-bar__icon"><!----></div> <div class="notifications-bar__content"><p>
          Hello world
          <!----></p></div></div><div class="notifications-bar notifications-bar--notification notifications-bar--visible"><div class="notifications-bar__icon"><!----></div> <div class="notifications-bar__content"><p>
          Welcome to
          <a href="props.notification.url">Homeday</a></p></div></div></span>
  <div class="notifications__sizer" style="height: 0px;"></div>
</div>
`;
